pkg/provider/oidc.go

```diff
@@ -0,0 +1,195 @@
+package provider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/metautils"
+	"github.com/zitadel/oidc/v3/pkg/client/rp"
+	"github.com/zitadel/oidc/v3/pkg/client/rs"
+	httphelper "github.com/zitadel/oidc/v3/pkg/http"
+	"github.com/zitadel/oidc/v3/pkg/oidc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/zitadel/zitadel-go/v2/pkg/info"
+)
+
+var (
+	ErrMissingHeader        = errors.New("auth header missing")
+	ErrInvalidHeader        = errors.New("invalid auth header")
+	ErrInvalidToken         = errors.New("invalid token")
+	ErrInvalidAuthorization = errors.New("invalid authorization")
+)
+
+func newRelayingPartyOIDC(ctx context.Context, oidc *configurationOIDC) (rp.RelyingParty, error) {
+	redirectURI := fmt.Sprintf("http://localhost:%v%v", oidc.port, oidc.callbackURL)
+	cookieHandler := httphelper.NewCookieHandler(key, key, httphelper.WithUnsecure())
+	options := []rp.Option{
+		rp.WithCookieHandler(cookieHandler),
+		rp.WithVerifierOpts(rp.WithIssuedAtOffset(5 * time.Second)),
+	}
+	if oidc.clientSecret == "" {
+		options = append(options, rp.WithPKCE(cookieHandler))
+	}
+	if oidc.keyPath != "" {
+		options = append(options, rp.WithJWTProfile(rp.SignerFromKeyPath(oidc.keyPath)))
+	}
+	return rp.NewRelyingPartyOIDC(ctx, oidc.issuer, oidc.clientID, oidc.clientSecret, redirectURI, oidc.scopes, options...)
+}
+
+func marshalUserinfo() rp.CodeExchangeUserinfoCallback[*oidc.IDTokenClaims, *oidc.UserInfo] {
+	return func(w http.ResponseWriter, r *http.Request, tokens *oidc.Tokens[*oidc.IDTokenClaims], state string, rp rp.RelyingParty, info *oidc.UserInfo) {
+		data, err := json.Marshal(info)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		w.Write(data)
+	}
+}
+
+func oidcAuthenticatedHTTPInterceptor(resourceServer rs.ResourceServer) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			ctx := req.Context()
+
+			token, err := checkOIDCTokenFromRequest(req)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusUnauthorized)
+				return
+			}
+
+			pi := info.ProviderInfoFromContext(ctx)
+			resp, err := checkIntrospectWithCached(ctx, resourceServer, token, pi.IntrospectionResponse)
```

Comment:
just for my understanding: how is this intended? and why is the introspection response always cached?

Reply:
BTW as it's calling the introspection endpoint, that would mean it's checking for authorization and not authentication

Reply:
it's only cached if an introspection response is already in the context, as for example the different interceptors are stacked. It's only calling the introspection endpoint to check if the token is still active, so you suggest to remove the introspection call and only check the token itself?